Selected locale should be consistent on load between Session and UI

caalador opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/flow/2857)
<!-- Reviewable:end -->

denis-anisimov commented:
Reviewed 5 of 10 files at r1, 8 of 8 files at r2.
Review status: 12 of 13 files reviewed at latest revision, 13 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKhx-yX8AWRbSfY6Wf:-KyKhx-yX8AWRbSfY6Wg:bmlvvkw) (no related file):*
Should we provide a default I18N provider?
Don't see it as a new class though there is some `DefaultProvider` mention in the code.
Should it be mentioned in the docs ?

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKkYDk4c3uADxlbVBD:-KyKkYDk4c3uADxlbVBE:b-fser03) (no related file):*
Please create a ticket for Spring boot ( as a Java application to handle `I18NPRovider`  interface).

---

*[flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc, line 16 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKgLE7iSNROm9gmgsq:-KyKgLE7iSNROm9gmgsr:bseubun) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc#L16)):*
> on

one ?

---

*[flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc, line 16 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKgNkDfHodnFwCdsx_:-KyKgNkDfHodnFwCdsxa:bdwiyn1) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc#L16)):*
> should be available.

must be ?

---

*[flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc, line 54 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKgzO2lKqe6QliOAhr:-KyKgzO2lKqe6QliOAhs:b-fl72yj) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc#L54)):*
> CacheBuilder
>             .newBuilder()

Is it thread safe ?

---

*[flow-server/src/main/java/com/vaadin/server/startup/I18NRegistryInitializer.java, line 63 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKifPKtvtzGHsTneLe:-KyKifPKtvtzGHsTneLf:b-87bvy9) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-server/src/main/java/com/vaadin/server/startup/I18NRegistryInitializer.java#L63)):*
> private static class DefaultProvider 

Is it temporary class until it's not implemented as a separate ticket or it's permanent?

I think we should provide some default provider which read e.g. single file once at startup.
Like Java `ResourceBundle`.

---

*[flow-server/src/main/java/com/vaadin/ui/i18n/I18NProvider.java, line 36 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKj6LwrOnI_be4OChL:-KyKj6LwrOnI_be4OChM:bg0i7up) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-server/src/main/java/com/vaadin/ui/i18n/I18NProvider.java#L36)):*
> return Arrays.asList(Locale.getDefault());

I would say it's totally useless provider which uses server default locale.
If application is deployed to the cloud then you most likely don't want to use it's default locale.
It's dangerous to allow implementor do not implement this method.
It should be abstract I'm pretty sure.

---

*[flow-server/src/main/java/com/vaadin/ui/i18n/I18NProvider.java, line 78 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKjau8TBBJcn0e_CPh:-KyKjau8TBBJcn0e_CPi:b22kno9) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-server/src/main/java/com/vaadin/ui/i18n/I18NProvider.java#L78)):*
> default Locale getLocale() {

I don't understand why it should returns the current locale.

What's the usecase?

It's SPI. It should not know anything about `UI` and etc.
It should only provide data being implemented by the developer.

Current locale is something that doesn't belong to SPI at all. It should be managed by `UI` or I18N API (_not_ `SPI`).

Or may be I don't understand the intention here.

---

*[flow-server/src/main/java/com/vaadin/util/LocaleUtil.java, line 42 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKk8oHQcb85P6vBLym:-KyKk8oHQcb85P6vBLyn:bkb2ig) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-server/src/main/java/com/vaadin/util/LocaleUtil.java#L42)):*
> Locale

`Optional<Locale>` ?

---

*[flow-server/src/main/java/com/vaadin/util/LocaleUtil.java, line 66 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKkCL_vbTcMS7oticx:-KyKkCL_vbTcMS7oticy:bkb2ig) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-server/src/main/java/com/vaadin/util/LocaleUtil.java#L66)):*
> Locale

`Optional<Locale>` ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc, line 16 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKgLE7iSNROm9gmgsq:-KyKvoyWNflJPC8zznqx:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc#L16)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> on

one ?
</blockquote></details>

Done.

---

*[flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc, line 16 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKgNkDfHodnFwCdsx_:-KyKvq-s2Pey7NWVePYJ:bmpdmdl) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc#L16)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> should be available.

must be ?
</blockquote></details>

No it's not required. We do not want to force anyone to use i18n. We just want to enable

---

*[flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc, line 54 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKgzO2lKqe6QliOAhr:-KyKwbHWJhWayFHbKHRv:b-3quu8y) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc#L54)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> CacheBuilder
>             .newBuilder()

Is it thread safe ?
</blockquote></details>

this should be static. It seems that I can have the code not in nested classes also. updated.

---

*[flow-server/src/main/java/com/vaadin/server/startup/I18NRegistryInitializer.java, line 63 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKifPKtvtzGHsTneLe:-KyKxJTUM1_0SXhlX03P:b-sfghem) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-server/src/main/java/com/vaadin/server/startup/I18NRegistryInitializer.java#L63)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> private static class DefaultProvider 

Is it temporary class until it's not implemented as a separate ticket or it's permanent?

I think we should provide some default provider which read e.g. single file once at startup.
Like Java `ResourceBundle`.
</blockquote></details>

Actually we don't need this at all here. Removed.

---

*[flow-server/src/main/java/com/vaadin/ui/i18n/I18NProvider.java, line 36 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKj6LwrOnI_be4OChL:-KyKxiiTO6U1BplkQ0lv:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-server/src/main/java/com/vaadin/ui/i18n/I18NProvider.java#L36)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> return Arrays.asList(Locale.getDefault());

I would say it's totally useless provider which uses server default locale.
If application is deployed to the cloud then you most likely don't want to use it's default locale.
It's dangerous to allow implementor do not implement this method.
It should be abstract I'm pretty sure.

</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/ui/i18n/I18NProvider.java, line 78 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKjau8TBBJcn0e_CPh:-KyKxnvLoayWc4KVSAHo:b-2mv16h) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-server/src/main/java/com/vaadin/ui/i18n/I18NProvider.java#L78)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> default Locale getLocale() {

I don't understand why it should returns the current locale.

What's the usecase?

It's SPI. It should not know anything about `UI` and etc.
It should only provide data being implemented by the developer.

Current locale is something that doesn't belong to SPI at all. It should be managed by `UI` or I18N API (_not_ `SPI`).

Or may be I don't understand the intention here.
</blockquote></details>

It was to make life easier, dropped and left for the developer to implement.

---

*[flow-server/src/main/java/com/vaadin/util/LocaleUtil.java, line 42 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKk8oHQcb85P6vBLym:-KyKyedqplXcFyZHV4M-:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-server/src/main/java/com/vaadin/util/LocaleUtil.java#L42)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> Locale

`Optional<Locale>` ?
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/util/LocaleUtil.java, line 66 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKkCL_vbTcMS7oticx:-KyKyf0YJ9PugSRBG3xx:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-server/src/main/java/com/vaadin/util/LocaleUtil.java#L66)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> Locale

`Optional<Locale>` ?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 8 of 9 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKhx-yX8AWRbSfY6Wf:-KyL7XKcTvVelP6Tz5FR:b-f05ok4) (no related file):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

We will not provide a default implementation except the one shown in the tutorials.
</blockquote></details>

Hm......
I would expect to have some.
Well, that another feature anyway. 
Can be created as a separate ticket.

---

*[flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc, line 54 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKgzO2lKqe6QliOAhr:-KyL7lnfHxDHtOKB_16H:b-epn2nt) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc#L54)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

this should be static. It seems that I can have the code not in nested classes also. updated.
</blockquote></details>

This is not what I meant.

I meant `LocaleCache` result which is the result of `CacheBuilder` execution.
It's an external class and I don't know how it's implemented.

But the `LoadingCache` instance is going to be used across different `UI`s and sessions.
It means that there will be race conditions for this resource.
As I understand `LoadingCache` uses lazy loading ( at least) and also has an expiration time.
It means that if it's access from the different threads then it will read/update/reload data from it's internal state.
This state has to be thread safe. If it 's not then it may fail.

This can be eliminated by the static load of the data right away with the class loading (without any lazy, expiration). So the data stays effectively final. 
Otherwise we should care about thread-safety.

T

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc, line 54 at r2](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyKgzO2lKqe6QliOAhr:-KyLA1ucUk-MmZVgCNlE:b292jmo) ([raw file](https://github.com/vaadin/flow/blob/04686314bf9332194fbe5814407bf06e944b49e4/flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc#L54)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

This is not what I meant.

I meant `LocaleCache` result which is the result of `CacheBuilder` execution.
It's an external class and I don't know how it's implemented.

But the `LoadingCache` instance is going to be used across different `UI`s and sessions.
It means that there will be race conditions for this resource.
As I understand `LoadingCache` uses lazy loading ( at least) and also has an expiration time.
It means that if it's access from the different threads then it will read/update/reload data from it's internal state.
This state has to be thread safe. If it 's not then it may fail.

This can be eliminated by the static load of the data right away with the class loading (without any lazy, expiration). So the data stays effectively final. 
Otherwise we should care about thread-safety.

T
</blockquote></details>

Implementations of this interface are expected to be thread-safe, and can be safely accessed by multiple concurrent threads.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyLZEZeiZbqN_Mobzs7:-KyLZEZfU_0uPiF0a06g:b-482y8g) (no related file):*
Lets hold this for a bit as I have another approach that I'll test for the initialization

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

caalador commented:
Updated with simpler and clearer usage.

---

Review status: 6 of 15 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857#-:-KyQ1KpQwvghxLuG0Nza:b4fdz2k)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 10 files at r4, 10 of 14 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyLZEZeiZbqN_Mobzs7:-KyQ2hgJvZjltu_MWER3:b-mczkoi) (no related file):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

Lets hold this for a bit as I have another approach that I'll test for the initialization
</blockquote></details>

I would propose to do it in some API defined way (even if it still uses the property).
In Spring Boot e.g. you don't create any servlet at all.
Instead it's created for you under the hood.

So it significantly complicate the way of specifying the I18N provider in this case.

I would suggest to use `Instantiator` API for this.
E.g. introduce an additional method which returns the provider.

Now user may provide a custom sublcass of `Instantiator` or he may uses a property in the implemented way.
The provided implementation of the method may uses the value from the property to provide a provider instance.

Spring has its own implementation of the `Instantiator`  class and it will make its own way to provide the provider instance.

Now you should then update Spring add-on to use one more property (see `SpringServlet::config`).
Well,  you should do it anyway if you want to keep it working via property.

---

*[flow-server/src/main/java/com/vaadin/server/I18NRegistry.java, line 27 at r7](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQ41ZlKDh6fmRuEq_V:-KyQ41ZlKDh6fmRuEq_W:bowtsmf) ([raw file](https://github.com/vaadin/flow/blob/4e8908fe7d7ae1c709528e30f0eef2a701d02b1e/flow-server/src/main/java/com/vaadin/server/I18NRegistry.java#L27)):*
> public class I18NRegistry {

Do you need this registry now at all ?

You can ask the provider on the fly and cache it somewhere once it's created (e.g. in `UI` or `UIInternals` or `VaadinSession` or whatever).
It's too cumbersome to use a dedicated registry now for this.

---

*[flow-server/src/test/java/com/vaadin/server/I18NRegistryTest.java, line 19 at r7](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQ4faPlbxhT5HRLIH4:-KyQ4faPlbxhT5HRLIH5:bxd3tly) ([raw file](https://github.com/vaadin/flow/blob/4e8908fe7d7ae1c709528e30f0eef2a701d02b1e/flow-server/src/test/java/com/vaadin/server/I18NRegistryTest.java#L19)):*
> /**
>  *
>  */

empty

---

*[flow-server/src/test/java/com/vaadin/server/TestProvider.java, line 19 at r7](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQ4iMhQmQR7oNcdeng:-KyQ4iMhQmQR7oNcdenh:bxd3tly) ([raw file](https://github.com/vaadin/flow/blob/4e8908fe7d7ae1c709528e30f0eef2a701d02b1e/flow-server/src/test/java/com/vaadin/server/TestProvider.java#L19)):*
> /**
>  *
>  */

empty

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

caalador commented:
Reviewed 1 of 14 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyLZEZeiZbqN_Mobzs7:-KyQJxUZqEWVUzeA-nsI:ban5fw9) (no related file):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

I would propose to do it in some API defined way (even if it still uses the property).
In Spring Boot e.g. you don't create any servlet at all.
Instead it's created for you under the hood.

So it significantly complicate the way of specifying the I18N provider in this case.

I would suggest to use `Instantiator` API for this.
E.g. introduce an additional method which returns the provider.

Now user may provide a custom sublcass of `Instantiator` or he may uses a property in the implemented way.
The provided implementation of the method may uses the value from the property to provide a provider instance.

Spring has its own implementation of the `Instantiator`  class and it will make its own way to provide the provider instance.

Now you should then update Spring add-on to use one more property (see `SpringServlet::config`).
Well,  you should do it anyway if you want to keep it working via property.
</blockquote></details>

The `Instantiator`is not usable as the first time we need the Provider if provided is during `createAndRegisterSession` which is before the bootstrap handle handles the request and creates a UI.

---

*[flow-server/src/main/java/com/vaadin/server/I18NRegistry.java, line 27 at r7](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQ41ZlKDh6fmRuEq_V:-KyQGgxmiGUjvnPx1pEd:b65oklg) ([raw file](https://github.com/vaadin/flow/blob/4e8908fe7d7ae1c709528e30f0eef2a701d02b1e/flow-server/src/main/java/com/vaadin/server/I18NRegistry.java#L27)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> public class I18NRegistry {

Do you need this registry now at all ?

You can ask the provider on the fly and cache it somewhere once it's created (e.g. in `UI` or `UIInternals` or `VaadinSession` or whatever).
It's too cumbersome to use a dedicated registry now for this.
</blockquote></details>

The i18n provider should be for all and not UI or Session specific.

---

*[flow-server/src/test/java/com/vaadin/server/I18NRegistryTest.java, line 19 at r7](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQ4faPlbxhT5HRLIH4:-KyQGqnQkL9NeauJI7zq:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/4e8908fe7d7ae1c709528e30f0eef2a701d02b1e/flow-server/src/test/java/com/vaadin/server/I18NRegistryTest.java#L19)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> /**
>  *
>  */

empty
</blockquote></details>

Done.

---

*[flow-server/src/test/java/com/vaadin/server/TestProvider.java, line 19 at r7](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQ4iMhQmQR7oNcdeng:-KyQHAtc_4PbTD4Kob2k:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/4e8908fe7d7ae1c709528e30f0eef2a701d02b1e/flow-server/src/test/java/com/vaadin/server/TestProvider.java#L19)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> /**
>  *
>  */

empty
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: 13 of 19 files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyLZEZeiZbqN_Mobzs7:-KyQMrgfteo3vbVqDXmE:b-gs0tf1) (no related file):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

The `Instantiator`is not usable as the first time we need the Provider if provided is during `createAndRegisterSession` which is before the bootstrap handle handles the request and creates a UI.
</blockquote></details>

I had a version where we checked `application.properties` for the parameter also.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: 10 of 24 files reviewed at latest revision, 5 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyLZEZeiZbqN_Mobzs7:-KyQm4lCvHOkxwNI1Taa:bei0oll) (no related file):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

I had a version where we checked `application.properties` for the parameter also.
</blockquote></details>

Now the initalization is done through the `Instantiator`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 12 of 14 files at r8, 1 of 2 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc, line 27 at r10](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQkKle83HsXj840kmi:-KyQkKlf3T_XvJxQR6nq:b-rv2xkk) ([raw file](https://github.com/vaadin/flow/blob/34d25ad1659be0f4036906ce6c39b15e6421d1a5/flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc#L27)):*
> Using `application.properties`, the properties file should be located in `src/main/resources` with the content:
>  

Does it always work or only for Spring ?
Every Spring property has a namespace: it's prefixed with "vaadin.".

OK, I've read the other code and I think it should be removed.

`application.properties` is strongly Spring related thing.
And it will work automatically if you keep the changes in `SpringServlet`.

---

*[flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java, line 97 at r10](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQn43x2tY-KGfK7B1D:-KyQn43x2tY-KGfK7B1E:bc3hixd) ([raw file](https://github.com/vaadin/flow/blob/34d25ad1659be0f4036906ce6c39b15e6421d1a5/flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java#L97)):*
> if (!I18NRegistry.getInstance().isInitialized()) {
>             getI18NProviderInstance()
>                     .ifPresent(I18NRegistry.getInstance()::setProvider);

It allows a race condition.

` if (!I18NRegistry.getInstance().isInitialized()) {`may be `true` but then it may become false when 
`I18NRegistry.getInstance()::setProvider` is called.

Depending on the original intention it may be an issue or not.

If you want to have only one I18N provider instance then it's an issue.

I think you should move this logic into the `I18NRegistry` class itself. 
Here method should only delegate to `I18NRegistry`.


`I18NRegistry` may have `final static` field which is initialized right away with the provider.
Such initialization is thread safe.

I'm not sure about necessity to have `setProvider` method at all and I'm not sure about `I18NRegistry` necessity  at all.

`Instantiator` is declared for `VaadinService` which is kind of singleton for the application.
And you most likely are not going to do anything until this singleton is created.
Once it's created you can store `Instantiator` instance somewhere for yourself to use I18N outside of UI (similarly to `I18NRegistry` but this will be you own responsibility). 
Most likely you don't need to use I18N provider outside your UI anyway.

So I would load the provider from the property (using the code below) in the static final field in the `Instantiator` and just return it here.
That's it.

---

*[flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java, line 98 at r10](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQmKvVkCdYw63efwC4:-KyQmKvVkCdYw63efwC5:b-l76uxr) ([raw file](https://github.com/vaadin/flow/blob/34d25ad1659be0f4036906ce6c39b15e6421d1a5/flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java#L98)):*
> I18NRegistry.getInstance().markInitialized();

Do you need this method separately ?

---

*[flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java, line 139 at r10](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQl3Sv1gQNvyLGb0kS:-KyQl3Sv1gQNvyLGb0kT:b-wtej2d) ([raw file](https://github.com/vaadin/flow/blob/34d25ad1659be0f4036906ce6c39b15e6421d1a5/flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java#L139)):*
> try (InputStream is = service.getClassLoader()
>                     .getResourceAsStream("application.properties")) {

No need to do this (just because Spring e.g.).
Properties are read from any `PropertySource`for Spring.
They are expected to be prefixed with "vaadin.".
Then those properties are set as init parameters.
Please see `SpringServlet::config`.


And you have already modified the `SpringServlet`.
So just read 
`VaadinSession.getCurrent().getConfiguration()
                .getStringProperty(Constants.I18N_PROVIDER, null); ` and that's it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc, line 27 at r10](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQkKle83HsXj840kmi:-KyUEuQ7AOv42ItTtCJt:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/34d25ad1659be0f4036906ce6c39b15e6421d1a5/flow-documentation/miscellaneous/tutorial-i18n-localization.asciidoc#L27)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> Using `application.properties`, the properties file should be located in `src/main/resources` with the content:
>  

Does it always work or only for Spring ?
Every Spring property has a namespace: it's prefixed with "vaadin.".

OK, I've read the other code and I think it should be removed.

`application.properties` is strongly Spring related thing.
And it will work automatically if you keep the changes in `SpringServlet`.
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java, line 97 at r10](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQn43x2tY-KGfK7B1D:-KyUJHwp8fQy9U7PF91Y:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/34d25ad1659be0f4036906ce6c39b15e6421d1a5/flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java#L97)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> if (!I18NRegistry.getInstance().isInitialized()) {
>             getI18NProviderInstance()
>                     .ifPresent(I18NRegistry.getInstance()::setProvider);

It allows a race condition.

` if (!I18NRegistry.getInstance().isInitialized()) {`may be `true` but then it may become false when 
`I18NRegistry.getInstance()::setProvider` is called.

Depending on the original intention it may be an issue or not.

If you want to have only one I18N provider instance then it's an issue.

I think you should move this logic into the `I18NRegistry` class itself. 
Here method should only delegate to `I18NRegistry`.


`I18NRegistry` may have `final static` field which is initialized right away with the provider.
Such initialization is thread safe.

I'm not sure about necessity to have `setProvider` method at all and I'm not sure about `I18NRegistry` necessity  at all.

`Instantiator` is declared for `VaadinService` which is kind of singleton for the application.
And you most likely are not going to do anything until this singleton is created.
Once it's created you can store `Instantiator` instance somewhere for yourself to use I18N outside of UI (similarly to `I18NRegistry` but this will be you own responsibility). 
Most likely you don't need to use I18N provider outside your UI anyway.

So I would load the provider from the property (using the code below) in the static final field in the `Instantiator` and just return it here.
That's it.
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java, line 98 at r10](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQmKvVkCdYw63efwC4:-KyUJIJQ_krCsIHFW_QU:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/34d25ad1659be0f4036906ce6c39b15e6421d1a5/flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java#L98)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> I18NRegistry.getInstance().markInitialized();

Do you need this method separately ?
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java, line 139 at r10](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQl3Sv1gQNvyLGb0kS:-KyUJOV3jTXwNL1eU3Vf:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/34d25ad1659be0f4036906ce6c39b15e6421d1a5/flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java#L139)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> try (InputStream is = service.getClassLoader()
>                     .getResourceAsStream("application.properties")) {

No need to do this (just because Spring e.g.).
Properties are read from any `PropertySource`for Spring.
They are expected to be prefixed with "vaadin.".
Then those properties are set as init parameters.
Please see `SpringServlet::config`.


And you have already modified the `SpringServlet`.
So just read 
`VaadinSession.getCurrent().getConfiguration()
                .getStringProperty(Constants.I18N_PROVIDER, null); ` and that's it.
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/I18NRegistry.java, line 27 at r7](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyQ41ZlKDh6fmRuEq_V:-KyUJQiVy3xgzd96JDbK:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/4e8908fe7d7ae1c709528e30f0eef2a701d02b1e/flow-server/src/main/java/com/vaadin/server/I18NRegistry.java#L27)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

The i18n provider should be for all and not UI or Session specific.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 9 of 9 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java, line 43 at r11](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyUUZY3WXKm9NlocEnQ:-KyUUZY4hWLWlVwrm6TV:b-u6p19g) ([raw file](https://github.com/vaadin/flow/blob/6aefe898451d0ede7a968340f9fc4d65e9778c13/flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java#L43)):*
> private static final AtomicReference<I18NProvider> i18nProvider = new AtomicReference<>();


If you want to do it in the CTOR (eagerly) then you don't need `AtomicReference`. Just make this field final and it's thread safe.

On the other hand it's not effective initialize the provider immediately in the CTOR.
It might be useful to load it lazily.

To make it lazy you should move this code into `getI18NProvider()` and check whether the field `i18nProvider`  has `null` as a value.
This still makes possible a race condition _BUT_ since you are using `compareAndSet`  then the new instance won't be set and used by anyone if race condition happens in unhappy path.
So the worst thing which may happen: one more provider instance will be created (but it won't be visible by anyone and will be GCed immediately).
Everything else is thread safe which is guaranteed by the `volatile`  (internally used by `AtomicReference`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: 26 of 27 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java, line 43 at r11](https://reviewable.io:443/reviews/vaadin/flow/2857#-KyUUZY3WXKm9NlocEnQ:-KyUZ-MvsF-aDZ4rCql6:bs7wuqo) ([raw file](https://github.com/vaadin/flow/blob/6aefe898451d0ede7a968340f9fc4d65e9778c13/flow-server/src/main/java/com/vaadin/flow/di/DefaultInstantiator.java#L43)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> private static final AtomicReference<I18NProvider> i18nProvider = new AtomicReference<>();


If you want to do it in the CTOR (eagerly) then you don't need `AtomicReference`. Just make this field final and it's thread safe.

On the other hand it's not effective initialize the provider immediately in the CTOR.
It might be useful to load it lazily.

To make it lazy you should move this code into `getI18NProvider()` and check whether the field `i18nProvider`  has `null` as a value.
This still makes possible a race condition _BUT_ since you are using `compareAndSet`  then the new instance won't be set and used by anyone if race condition happens in unhappy path.
So the worst thing which may happen: one more provider instance will be created (but it won't be visible by anyone and will be GCed immediately).
Everything else is thread safe which is guaranteed by the `volatile`  (internally used by `AtomicReference`).
</blockquote></details>

Done.§

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2857)*
<!-- Sent from Reviewable.io -->
